nomos-services/key-management-system-macros/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+    ident: &Ident,
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let key_enum_impl_secured_key_method_sign =
+        build_key_enum_impl_secured_key_method_sign(key_enum_variants);
+    let key_enum_impl_secured_key_method_as_public_key =
+        build_key_enum_impl_secured_key_method_as_public_key(key_enum_variants);
+
+    quote! {
+        impl crate::keys::secured_key::SecuredKey for #ident
+        {
+            type Payload = PayloadEncoding;
+            type Signature = SignatureEncoding;
+            type PublicKey = PublicKeyEncoding;
+            type Error = crate::keys::errors::KeyError;
+
+            #key_enum_impl_secured_key_method_sign
+            #key_enum_impl_secured_key_method_as_public_key
+        }
+    }
+}
+
+fn build_key_enum_impl_secured_key_method_sign(
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let match_block = if key_enum_variants.is_empty() {
+        quote! {
+            Err(Self::Error::NoKeysEnabled)
+        }
+    } else {
+        let sign_arms_ok = key_enum_variants.iter().map(|variant| {
+            let variant_ident = &variant.ident;
+            let variant_attributes_cfg: Vec<Attribute> =
+                get_cfg_attributes(&variant.attrs).collect();
+
+            quote! {
+                #(#variant_attributes_cfg)*
+                (Self::#variant_ident(key), Self::Payload::#variant_ident(payload)) => {
+                    key.sign(payload).map(Self::Signature::#variant_ident)
+                }
+            }
+        });
+
+        let sign_arm_error = quote! {
+            (key, payload) => Err(crate::keys::errors::EncodingError::requires(key, payload).into()),
+        };
+
+        quote! {
+            match (self, payload) {
+                #(#sign_arms_ok)*
+                #sign_arm_error
+            }
+        }
+    };
+
+    quote! {
+        fn sign(&self, payload: &Self::Payload) -> Result<Self::Signature, Self::Error> {
+            #match_block
+        }
+    }
+}
+
+fn build_key_enum_impl_secured_key_method_as_public_key(
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let match_block = if key_enum_variants.is_empty() {
+        quote! {
+            Err(Self::Error::NoKeysEnabled)
+        }
+    } else {
+        let as_public_key_arms = key_enum_variants.iter().map(|variant| {
+            let variant_ident = &variant.ident;
+            let variant_attributes_cfg: Vec<Attribute> = get_cfg_attributes(&variant.attrs).collect();
+            quote! {
+                #(#variant_attributes_cfg)*
+                Self::#variant_ident(key) => key.as_public_key().map(Self::PublicKey::#variant_ident),
+            }
+        });
+
+        quote! {
+            match self {
+                #(#as_public_key_arms)*
+            }
+        }
+    };
+
+    quote! {
+        fn as_public_key(&self) -> Result<Self::PublicKey, Self::Error> {
+            #match_block
+        }
+    }
```

Comment:
So, we changed the api of the trait to support empty enums. IMO is completely unnecesary. I would rather panic on empty enum so the compilation fails.

Reply:
I'll give the full answer in the thread above.